### PR TITLE
Upgrade to Bazel 4.0

### PR DESCRIPTION
**Pull request by cocreature**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -33,6 +33,9 @@ build --experimental_strict_action_env
 # in us getting a certificate for githubassets.com instead.
 startup --host_jvm_args=-Xms1g --host_jvm_args=-Xmx1g --host_jvm_args=-Djdk.tls.client.protocols=TLSv1.2
 
+# Enable sandboxing and caching for exclusive tests
```
Comment (cocreature) [inline .bazelrc:36]: this is the equivalent of the patch we had before. Nice to see this finally upstreamed.

File: bazel_tools/rules-nixpkgs-llvm-cov.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/nixpkgs/toolchains/cc.nix b/nixpkgs/toolchains/cc.nix
```
Comment (cocreature) [inline bazel_tools/rules-nixpkgs-llvm-cov.patch:1]: Already upstreamed but I’ll keep a rules-nixpkgs upgrade for a separate PR.

File: compatibility/deps.bzl [hunk 3]
```diff
@@ -72,6 +72,16 @@ def daml_deps():
...
             urls = ["https://github.com/tweag/rules_nixpkgs/archive/%s.tar.gz" % rules_nixpkgs_version],
             sha256 = rules_nixpkgs_sha256,
+            patches = [
+                # On CI and locally we observe occasional segmantation faults
+                # of nix. A known issue since Nix 2.2.2 is that HTTP2 support
+                # can cause such segmentation faults. Since Nix 2.3.2 it is
+                # possible to disable HTTP2 via a command-line flag, which
+                # reportedly solves the issue. See
+                # https://github.com/NixOS/nix/issues/2733#issuecomment-518324335
+                "@daml//bazel_tools:nixpkgs-disable-http2.patch",
```
Comment (cocreature) [inline compatibility/deps.bzl:82]: I don’t think we strictly need this but while I get patches in sync between main and compatibility, I might as well sync up both.

File: deps.bzl [hunk 4]
```diff
@@ -358,14 +359,14 @@ genrule(
...
     $(execpath @jekyll_nix//:bin/jekyll) build -s $$DIR
-    tar hc _site \\
-        --owner=1000 \\
-        --group=1000 \\
-        --mtime=2000-01-01\\ 00:00Z \\
-        --no-acls \\
-        --no-xattrs \\
-        --no-selinux \\
-        --sort=name \\
+    tar hc _site \\\\
```
Comment (cocreature) [inline deps.bzl:363]: This was an invalid escape sequence before which Bazel implicitly translated to this. Now it errors out on this.

File: nix/bazel-cc-wrapper-response-file.patch [hunk 5]
```diff
@@ -0,0 +1,32 @@
+diff --git a/tools/cpp/osx_cc_wrapper.sh.tpl b/tools/cpp/osx_cc_wrapper.sh.tpl
```
Comment (cocreature) [inline nix/bazel-cc-wrapper-response-file.patch:1]: should be upstreamed to Bazel but for now this gets us going.

File: ci/build.yml [hunk 6]
```diff
@@ -40,7 +40,7 @@ jobs:
   timeoutInMinutes: 360
   pool:
     name: 'ubuntu_20_04'
-    demands: assignment -equals default
+    demands: assignment -equals nixpkgs-upgrade
```
Comment (cocreature) [inline ci/build.yml:43]: will remove all those before merging.

**Review by aherrmann-da — APPROVED** [review]
Nice, thank you!

**Pull request merged**